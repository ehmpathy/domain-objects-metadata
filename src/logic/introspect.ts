import ts from 'typescript';

import { DomainObjectMetadata } from '../domain';
import { extractRelevantProgramSourceFiles } from './extractRelevantProgramSourceFiles';
import { getHydratedDomainObjectMetadatasFromFiles } from './hydrate/getHydratedDomainObjectMetadatasFromFiles';

export const introspect = (
  filePaths: string | string[],
): DomainObjectMetadata[] => {
  // normalize the input
  const filePathsToIntrospect = Array.isArray(filePaths)
    ? filePaths
    : [filePaths]; // normalize to array, since its the most general case

  // create a "program" instance: a collection of source files
  const program = ts.createProgram(filePathsToIntrospect, {});

  // filter out irrelevant source files
  const files = extractRelevantProgramSourceFiles(program.getSourceFiles());

  // now get hydrated domain object metadatas from that file
  return getHydratedDomainObjectMetadatasFromFiles(files);
};
