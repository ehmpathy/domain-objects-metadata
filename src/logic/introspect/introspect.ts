/* eslint-disable no-restricted-syntax */
import * as ts from 'typescript';

type DomainObjectMetadata = any;

const nodeIsADomainObjectClass = (node: ts.Node): node is ts.ClassDeclaration & { name: string } => {
  // if this node does not reference a domain-object class, return nothing
  if (!ts.isClassDeclaration(node)) return false; // not a class, not of interest
  if (!node.name) return false; // no name, not of interest
  if (!node.heritageClauses?.some((clause) => clause.token === ts.SyntaxKind.ExtendsKeyword)) return false; // does not extend anything, not of interest
  console.log(node.heritageClauses);
  return true;
};

// const extractDomainObjectMetadataFromSourceFileNode = ({ node, checker }: { node: ts.Node, checker: ts.TypeChecker }) => {
//   // if this is a namespace, visit its children
//   if (!nodeIsADomainObjectClass(node))

//   // if it is a domain object though, grab its deets
//     let symbol = checker.getSymbolAtLocation(node.name);
//     if (symbol) {
//       output.push(serializeClass(symbol));
//     }
//     // No need to walk any further, class expressions/inner declarations
//     // cannot be exported
//   } else if (ts.isModuleDeclaration(node)) {
//     // This is a namespace, visit its children
//     ts.forEachChild(node, visit);
//   }
// };

export const introspect = (filePaths: string | string[]) => {
  // normalize the input
  const filePathsToIntrospect = Array.isArray(filePaths) ? filePaths : [filePaths]; // normalize to array, since its the most general case

  // create a "program" instance: a collection of source files
  const program = ts.createProgram(filePathsToIntrospect, {});

  // Get the checker, we will use it to find more about classes
  const checker = program.getTypeChecker();
  const output: DomainObjectMetadata[] = [];

  // for each non-node-module file in the program, go through it and find all of the DomainObject definitions

  // visit every node in the files to define the domain object class definition nodes
  const domainObjectClassDefinitionNodes: ts.Node[] = [];
  program.getSourceFiles().forEach((sourceFile) => {
    console.log(Object.keys(sourceFile));
  });
  // for (const sourceFile of program.getSourceFiles()) {
  //   console.log(sourceFile);
  //   if (!sourceFile.isDeclarationFile) {
  //     // walk the tree to search for domain-objects
  //     ts.forEachChild(sourceFile, (node) => {
  //       const metadata = extractDomainObjectMetadataFromSourceFileNode({ node });
  //     });
  //   }
  // }

  // // Visit every sourceFile in the program
  // for (const sourceFile of program.getSourceFiles()) {
  //   if (!sourceFile.isDeclarationFile) {
  //     // walk the tree to search for domain-objects
  //     ts.forEachChild(sourceFile, (node) => {
  //       const metadata = extractDomainObjectMetadataFromSourceFileNode({ node });
  //     });
  //   }
  // }
  // print out the doc
  console.log(JSON.stringify(output, undefined, 2));
};
