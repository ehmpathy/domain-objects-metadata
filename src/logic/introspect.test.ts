import { introspect } from './introspect';

describe('introspect', () => {
  it('should be possible to introspect the Address file', () => {
    const metadatas = introspect(`${__dirname}/__test_assets__/Address.ts`);
    // console.log(JSON.stringify(metadatas, null, 2));
    expect(metadatas.length).toEqual(1);
    expect(metadatas).toMatchSnapshot();
  });
  it('should be possible to introspect the DeliveryVan file', () => {
    const metadatas = introspect(`${__dirname}/__test_assets__/DeliveryVan.ts`);
    // console.log(JSON.stringify(metadatas, null, 2));
    expect(metadatas.length).toEqual(1);
    expect(metadatas).toMatchSnapshot();
  });
  it('should be possible to introspect the Delivery file', () => {
    const metadatas = introspect(`${__dirname}/__test_assets__/Delivery.ts`);
    // console.log(JSON.stringify(metadatas, null, 2));
    expect(metadatas.length).toEqual(4);
    expect(metadatas).toMatchSnapshot();
  });
  it('should be possible to introspect all of those files at the same time', () => {
    const metadatas = introspect([
      `${__dirname}/__test_assets__/Delivery.ts`,
      `${__dirname}/__test_assets__/Address.ts`,
      `${__dirname}/__test_assets__/DeliveryVan.ts`,
    ]);
    // console.log(JSON.stringify(metadatas, null, 2));
    expect(metadatas.length).toEqual(4);
    expect(metadatas).toMatchSnapshot();
  });
});
