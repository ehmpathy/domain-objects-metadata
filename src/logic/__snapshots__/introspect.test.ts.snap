// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`introspect should be possible to introspect all of those files at the same time 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": null,
      "updatable": null,
    },
    "extends": "DomainLiteral",
    "name": "Address",
    "properties": {
      "city": DomainObjectPropertyMetadata {
        "name": "city",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "postal": DomainObjectPropertyMetadata {
        "name": "postal",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "state": DomainObjectPropertyMetadata {
        "name": "state",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "street": DomainObjectPropertyMetadata {
        "name": "street",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "suite": DomainObjectPropertyMetadata {
        "name": "suite",
        "nullable": true,
        "required": true,
        "type": "STRING",
      },
    },
  },
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "vin",
      ],
      "updatable": [
        "milage",
      ],
    },
    "extends": "DomainEntity",
    "name": "DeliveryVan",
    "properties": {
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "make": DomainObjectPropertyMetadata {
        "name": "make",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "milage": DomainObjectPropertyMetadata {
        "name": "milage",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "model": DomainObjectPropertyMetadata {
        "name": "model",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
      "vin": DomainObjectPropertyMetadata {
        "name": "vin",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "year": DomainObjectPropertyMetadata {
        "name": "year",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
    },
  },
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "trackingNumber",
      ],
      "updatable": null,
    },
    "extends": "DomainEntity",
    "name": "Package",
    "properties": {
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "trackingNumber": DomainObjectPropertyMetadata {
        "name": "trackingNumber",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
      "weight": DomainObjectPropertyMetadata {
        "name": "weight",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
    },
  },
  DomainObjectMetadata {
    "decorations": {
      "unique": [],
      "updatable": [
        "status",
        "contactInfo",
      ],
    },
    "extends": "DomainEntity",
    "name": "Delivery",
    "properties": {
      "contactInfo": DomainObjectPropertyMetadata {
        "name": "contactInfo",
        "nullable": false,
        "of": {
          "nullable": false,
          "required": true,
          "type": "STRING",
        },
        "required": true,
        "type": "ARRAY",
      },
      "destination": DomainObjectPropertyMetadata {
        "name": "destination",
        "nullable": false,
        "of": DomainObjectReferenceMetadata {
          "extends": "DomainLiteral",
          "name": "Address",
        },
        "required": true,
        "type": "REFERENCE",
      },
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "packages": DomainObjectPropertyMetadata {
        "name": "packages",
        "nullable": false,
        "of": {
          "nullable": false,
          "of": DomainObjectReferenceMetadata {
            "extends": "DomainEntity",
            "name": "Package",
          },
          "required": true,
          "type": "REFERENCE",
        },
        "required": true,
        "type": "ARRAY",
      },
      "status": DomainObjectPropertyMetadata {
        "name": "status",
        "nullable": false,
        "of": [
          "SCHEDULED",
          "IN_PROGRESS",
          "DELIVERED",
          "FAILED",
          "CANCELED",
        ],
        "required": true,
        "type": "ENUM",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
      "van": DomainObjectPropertyMetadata {
        "name": "van",
        "nullable": false,
        "of": DomainObjectReferenceMetadata {
          "extends": "DomainEntity",
          "name": "DeliveryVan",
        },
        "required": true,
        "type": "REFERENCE",
      },
    },
  },
]
`;

exports[`introspect should be possible to introspect the Address file 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": null,
      "updatable": null,
    },
    "extends": "DomainLiteral",
    "name": "Address",
    "properties": {
      "city": DomainObjectPropertyMetadata {
        "name": "city",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "postal": DomainObjectPropertyMetadata {
        "name": "postal",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "state": DomainObjectPropertyMetadata {
        "name": "state",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "street": DomainObjectPropertyMetadata {
        "name": "street",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "suite": DomainObjectPropertyMetadata {
        "name": "suite",
        "nullable": true,
        "required": true,
        "type": "STRING",
      },
    },
  },
]
`;

exports[`introspect should be possible to introspect the Delivery file 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": null,
      "updatable": null,
    },
    "extends": "DomainLiteral",
    "name": "Address",
    "properties": {
      "city": DomainObjectPropertyMetadata {
        "name": "city",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "postal": DomainObjectPropertyMetadata {
        "name": "postal",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "state": DomainObjectPropertyMetadata {
        "name": "state",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "street": DomainObjectPropertyMetadata {
        "name": "street",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "suite": DomainObjectPropertyMetadata {
        "name": "suite",
        "nullable": true,
        "required": true,
        "type": "STRING",
      },
    },
  },
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "vin",
      ],
      "updatable": [
        "milage",
      ],
    },
    "extends": "DomainEntity",
    "name": "DeliveryVan",
    "properties": {
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "make": DomainObjectPropertyMetadata {
        "name": "make",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "milage": DomainObjectPropertyMetadata {
        "name": "milage",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "model": DomainObjectPropertyMetadata {
        "name": "model",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
      "vin": DomainObjectPropertyMetadata {
        "name": "vin",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "year": DomainObjectPropertyMetadata {
        "name": "year",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
    },
  },
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "trackingNumber",
      ],
      "updatable": null,
    },
    "extends": "DomainEntity",
    "name": "Package",
    "properties": {
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "trackingNumber": DomainObjectPropertyMetadata {
        "name": "trackingNumber",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
      "weight": DomainObjectPropertyMetadata {
        "name": "weight",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
    },
  },
  DomainObjectMetadata {
    "decorations": {
      "unique": [],
      "updatable": [
        "status",
        "contactInfo",
      ],
    },
    "extends": "DomainEntity",
    "name": "Delivery",
    "properties": {
      "contactInfo": DomainObjectPropertyMetadata {
        "name": "contactInfo",
        "nullable": false,
        "of": {
          "nullable": false,
          "required": true,
          "type": "STRING",
        },
        "required": true,
        "type": "ARRAY",
      },
      "destination": DomainObjectPropertyMetadata {
        "name": "destination",
        "nullable": false,
        "of": DomainObjectReferenceMetadata {
          "extends": "DomainLiteral",
          "name": "Address",
        },
        "required": true,
        "type": "REFERENCE",
      },
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "packages": DomainObjectPropertyMetadata {
        "name": "packages",
        "nullable": false,
        "of": {
          "nullable": false,
          "of": DomainObjectReferenceMetadata {
            "extends": "DomainEntity",
            "name": "Package",
          },
          "required": true,
          "type": "REFERENCE",
        },
        "required": true,
        "type": "ARRAY",
      },
      "status": DomainObjectPropertyMetadata {
        "name": "status",
        "nullable": false,
        "of": [
          "SCHEDULED",
          "IN_PROGRESS",
          "DELIVERED",
          "FAILED",
          "CANCELED",
        ],
        "required": true,
        "type": "ENUM",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
      "van": DomainObjectPropertyMetadata {
        "name": "van",
        "nullable": false,
        "of": DomainObjectReferenceMetadata {
          "extends": "DomainEntity",
          "name": "DeliveryVan",
        },
        "required": true,
        "type": "REFERENCE",
      },
    },
  },
]
`;

exports[`introspect should be possible to introspect the DeliveryVan file 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "vin",
      ],
      "updatable": [
        "milage",
      ],
    },
    "extends": "DomainEntity",
    "name": "DeliveryVan",
    "properties": {
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "make": DomainObjectPropertyMetadata {
        "name": "make",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "milage": DomainObjectPropertyMetadata {
        "name": "milage",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "model": DomainObjectPropertyMetadata {
        "name": "model",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
      "vin": DomainObjectPropertyMetadata {
        "name": "vin",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "year": DomainObjectPropertyMetadata {
        "name": "year",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
    },
  },
]
`;
